firmware/library/L1_Drivers/can.hpp

```diff
@@ -1,49 +1,828 @@
+    // Queue lengths
+    inline static constexpr uint8_t kTxQueueLength = kTxLength;
+    inline static constexpr uint8_t kRxQueueLength = kRxLength;
+
+    // Queue item sizes
+    inline static constexpr uint16_t kTxQueueItemSize = sizeof(kTxItem);
+    inline static constexpr uint16_t kRxQueueItemSize = sizeof(kRxItem);
+
+    // Queue storage area
+    inline static uint8_t
+        tx_queue_storage_area[kTxQueueLength * kTxQueueItemSize];
+    inline static uint8_t
+        rx_queue_storage_area[kRxQueueLength * kRxQueueItemSize];
   };
 
-  virtual bool Initialize()                       = 0;
-  virtual bool Send(const CanMessage_t & message) = 0;
-  virtual bool Receive(CanMessage_t * message)    = 0;
-  virtual bool SelfTestBus(const TestType kType)  = 0;
-  virtual bool IsBusOff()                         = 0;
-  virtual void EnableBus()                        = 0;
+  enum Ports : uint8_t
+  {
+    // based on schematic rev1.D
+    kCan1Port = 0,
+    kCan2Port = 2
+  };
+
+  enum Pins : uint8_t
+  {
+    // based on schematic rev1.D
+    kRd1PinNumber = 0,
+    kTd1PinNumber = 1,
+    kRd2PinNumber = 7,
+    kTd2PinNumber = 8,
+  };
+
+  enum PinFunctions : uint8_t
+  {
+    kRd1FunctionBit = 1,
+    kTd1FunctionBit = 1,
+    kRd2FunctionBit = 1,
+    kTd2FunctionBit = 1,
+  };
+
+  enum Interrupts : uint8_t
+  {
+    kRxBufferIntEnableBit  = 0,
+    kTxBuffer1IntEnableBit = 1,
+    kTxBuffer2IntEnableBit = 9,
+    kTxBuffer3IntEnableBit = 10,
+  };
+
+  enum Modes : uint8_t
+  {
+    kReset      = 0,
+    kListenOnly = 1,
+    kSelfTest   = 2,
+    kTxPriority = 3,
+    kSleepMode  = 4,
+    kRxPolarity = 5,
+    kTest       = 7,
+  };
+
+  // TODO(rabeelio): Add more baud rates
+  enum BaudRates : uint8_t
+  {
+    kBaud100Kbps = 100,
+  };
+
+  enum TxBuffers : uint8_t
+  {
+    kBuffer1 = 0,
+    kBuffer2 = 1,
+    kBuffer3 = 2,
+  };
+
+  // https://www.nxp.com/docs/en/user-guide/UM10562.pdf (pg. 554)
+  enum Commands : uint8_t
+  {
+    kReleaseRxBuffer            = 0x04,
+    kSendTxBuffer1              = 0x21,
+    kSendTxBuffer2              = 0x41,
+    kSendTxBuffer3              = 0x81,
+    kSelfReceptionSendTxBuffer1 = 0x30,
+    kAcceptAllMessages          = 0x02,
+  };
+
+  // Default constructor that defaults to CAN 1
+  constexpr Can()
+      : controller_(kCan1),
+        baud_rate_(BaudRates::kBaud100Kbps),
+        rd_(&rd_pin_),
+        td_(&td_pin_),
+        rd_pin_(Pin(kCan1Port, kRd1PinNumber)),
+        td_pin_(Pin(kCan1Port, kTd1PinNumber))
+  {
+  }
+
+  constexpr Can(Controllers controller, BaudRates baud_rate,
+                PinInterface * td_pin, PinInterface * rd_pin)
+      : controller_(controller),
+        baud_rate_(baud_rate),
+        rd_(td_pin),
+        td_(rd_pin),
+        rd_pin_(Pin::CreateInactivePin()),
+        td_pin_(Pin::CreateInactivePin())
+  {
+  }
+
+  bool Initialize() override
+  {
+    bool success = 1;
+    if (controller_ > kNumberOfControllers)
+    {
+      success = 0;
```

Comment:
```suggestion
      success = false;
```
Use true and false with bools.

Reply:
done